### ISSUE TITLE
403 forbidden for jupyter-server-proxy cross-origin

### ISSUE THREAD

**Issue by fonsp** [fonsp/pluto-on-binder#9]
_Posting an issue to record this for future googlers_

This took me a long time to fix: we use jupyter-server-proxy to run Pluto (custom IDE) on binder, which works fine. But for this project all requests need to be done cross-origin (very similar [thebe](https://github.com/executablebooks/thebe)).

By setting the header `acess-allow-cross-origin` to `*` on our server, all HTTP requests work (jupyter-server-proxy and binder ingress don't change this header), but **a websocket connection still failed, with a 403 response**.

**Comment by fonsp** [comment]
The cause was: `jupyter-server-proxy` uses the [tornado](https://github.com/tornadoweb/tornado) package for their websocket forwarding, but tornado 4 got a 'feature' where it only allows websocket connections if the origin and host are the same. _(WebSocket connections normally bypass CORS policy, which is a common security issue, which the tornado authors solved.)_

This means that the websocket connection (HTTP Upgrade request) never reached our server, and tornado wrote the 403 forbidden response.

---


# Solution

This SO answer is our solution:

https://stackoverflow.com/questions/24800436/under-tornado-v4-websocket-connections-get-refused-with-403

But how do we apply it?

## Long version

We need to fork `jupyter-server-proxy`, and make the code change. The code change is here:

https://github.com/fonsp/jupyter-server-proxy/commit/06cd1d843b7ada31d719272ae76d3c9bb777797a

The next tricky bit is installing this fork of jupyter-server-proxy instead of the pip-registered version in our binder.

My suggestion is: **tag your branch with a new, strange version number**. In my case, `1.5.1002` instead of `1.5.2`. I do that here: https://github.com/fonsp/jupyter-server-proxy/commit/3a58aa5005f942d0c208eab9a480f6ab171142ef

Next, use this exact version as your dependency in binder, that way the build fails if it's not using your fork. 

## Short version

The fork is here:
https://github.com/fonsp/jupyter-server-proxy

Commit SHA:
`3a58aa5005f942d0c208eab9a480f6ab171142ef`

You need to update `requirements.txt` or `environment.yml`: delete the pip version, add your fork. And ALSO update your setup.py. Final result here, search for every occurance of jupyter-server-proxy

https://github.com/fonsp/pluto-on-binder/tree/c8eb7a8ed1626b18ef2fec8002c1413ef9f75fa0
